### ISSUE TITLE
dealing with nested sentences

### ISSUE THREAD

**Issue by mreinstein** [spencermountain/compromise#935]
I'm evaluating compromise for building a scripting/narration engine for a project. I'm trying to build a command to make a character say something, and the quotes within a single sentence are giving me problems. For example:

```javascript
const input = `The girl says "Hey! gimme the meatballs!".`
const sentences = nlp(input).json().map(o=> o.text)
console.log('sentences:', sentences)     // prints   ['The girl says "Hey!', 'gimme the meatballs!".']
```

I was expecting this to be one sentence.  Any recommendations on how to deal with these types of input? My plan was to split the entire input text up into sentences and then run a small set of matches against each sentence like `nlp(sentence).match('#Noun? says "[.]")` but this won't work if a sentence with quotes is getting split out.

**Comment by spencermountain** [comment]
You're right. I always struggle with this, personally.

I mean, if the quote spans 10 sentences, up to some point, .sentences() should tokenize them, but with examples like yours, its clear. They are couched in the containing sentence.

Is there a clear best-practice rule of thumb somewhere with this?

I'm happy to improve this. A solution may be some weird config, or some magic length cutoff. Best though, if there is was a stronger understanding, on my end.

Maybe if there is a closing period *after* the quotation?
Is that he trick? The quote closes, and it doesn't look like the beginning of a new full sentence?

```
Before "one two three. Four five six." after.
Then second sentence.
```

```
Before "one two three. Four five six. Seven eight nine."
Then fourth sentence.
```

I dunno.

**Comment by mreinstein** [comment]
I guess part of the problem is `nlp(input).json()` assumes no hierarchical/nesting; it is purely a flat list.  So even if we came up with some rules for finding fully couched sentences, the existing data structure wouldn't represent it nicely.

I'm also open to other suggestions one could use to declare spoken text.  This was the first syntax that came to mind:
```
The girl says "You idiot, you're trapped again!". 
```

but I'd happily entertain other formats to for indicating an actor is speaking.


My current workaround is pre-parsing the raw input text, substituting sentence ending punctuation chars for some ascii symbols that I know will never be used in "real" sentences. Then I pass them through compromise, and then replace back:

```javascript
function getSentences (input) {
    let inString = false
    let result = ''

    for (let i=0; i < input.length; i++) {
        let c = input[i]

        if (inString) {
            if (c === '"' || c === '\'') {
                inString = false
            } else {
                if (c === '!')
                    c = '½'
                else if (c === '?')
                    c = '¾'
                else if (c === '.')
                    c = '¼'
            }
        } else if (c === '"' || c === '\'') {
            inString = true
        }

        result += c
    }

    return nlp(result).json().map(o => o.text.replaceAll('½', '!').replaceAll('¾', '?').replaceAll('¼', '.'))
}

const sentences = getSentences(input)

console.log('sentences:', sentences)
```

By the way thanks for this library, and for your consistent, steady maintenance! It's really an incredible thing you've come up with. ❤️

**Comment by spencermountain** [comment]
added support for short embedded sentences in quotes and parentheses in `14.4.1`.
arbitrary max quote length is 280 chars - let me know if you run into any problems
cheers

**Comment by mreinstein** [comment]
wow, fantastic!!

**Comment by mreinstein** [comment]
@spencermountain  maybe I'm using it wrong, but these results are unexpected:
```javascript
const input = `The hero was stunned by the scary monster. The glowing girl said "Hey! Leave him alone!".`

const sentences = nlp(input).json()
                            .map(o => o.text)

console.log('sentences:', sentences)
```

prints out `sentences: ['The hero was stunned by the scary monster.']`

**Comment by spencermountain** [comment]
whoa! this is bad!!
thank you for letting me know. will take a look at this asap

**Comment by spencermountain** [comment]
should be fixed in `14.4.2`
thanks!

**Comment by mreinstein** [comment]
Confirmed, seems to work for a variety of cases I've tried.

Thanks again for this features, it's freakin' sweet!
